src/Microsoft.VisualStudio.Threading/AwaitExtensions.cs

```diff
@@ -442,6 +456,71 @@ public void GetResult()
             }
         }
 
+        /// <summary>
+        /// An awaiter returned from <see cref="GetAwaiter(SynchronizationContext)"/>.
+        /// </summary>
+        public readonly struct SynchronizationContextAwaiter : ICriticalNotifyCompletion
+        {
+            private static readonly SendOrPostCallback SyncContextDelegate = s => ((Action)s!)();
+
+            /// <summary>
+            /// The context for continuations.
+            /// </summary>
+            private readonly SynchronizationContext syncContext;
+
+            /// <summary>
+            /// Initializes a new instance of the <see cref="SynchronizationContextAwaiter"/> struct.
+            /// </summary>
+            /// <param name="syncContext">The context for continuations.</param>
+            public SynchronizationContextAwaiter(SynchronizationContext syncContext)
+            {
+                this.syncContext = syncContext;
+            }
+
+            /// <summary>
+            /// Gets a value indicating whether no yield is necessary.
+            /// </summary>
+            /// <value>Always returns <see langword="false"/>.</value>
+            public bool IsCompleted => false;
```

Comment:
Do we ever want to short-cut this awaiter if the current SynchorizationContext matches what we want here?